executor/window.go

```diff
@@ -30,137 +30,133 @@ import (
+		e.remainingRowsInChunk = e.remainingRowsInChunk[1:]
 	}
 	return nil
 }
 
-func (e *WindowExec) consumeOneGroup(ctx context.Context, chk *chunk.Chunk) error {
-	var err error
-	if err = e.fetchChildIfNecessary(ctx, chk); err != nil {
-		return errors.Trace(err)
-	}
-	for ; e.inputRow != e.inputIter.End(); e.inputRow = e.inputIter.Next() {
-		e.meetNewGroup, err = e.groupChecker.meetNewGroup(e.inputRow)
+func (e *WindowExec) preparedChunkAvailable() bool {
+	return len(e.resultChunks) > 0 && e.remainingRowsInChunk[0] == 0
+}
+
+func (e *WindowExec) consumeOneGroup(ctx context.Context) error {
+	var groupRows []chunk.Row
+	for {
+		eof, err := e.fetchChildIfNecessary(ctx)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		if e.meetNewGroup && e.remainingRowsInGroup > 0 {
-			err := e.consumeGroupRows()
+		if eof {
+			e.executed = true
+			return e.consumeGroupRows(groupRows)
+		}
+		for inputRow := e.inputIter.Current(); inputRow != e.inputIter.End(); inputRow = e.inputIter.Next() {
+			meetNewGroup, err := e.groupChecker.meetNewGroup(inputRow)
 			if err != nil {
 				return errors.Trace(err)
 			}
-			err = e.appendResult2Chunk(chk)
-			return err
+			if meetNewGroup {
+				return e.consumeGroupRows(groupRows)
+			}
+			groupRows = append(groupRows, inputRow)
 		}
-		e.remainingRowsInGroup++
-		e.groupRows = append(e.groupRows, e.inputRow)
 	}
-	return nil
 }
 
-func (e *WindowExec) consumeGroupRows() (err error) {
-	if len(e.groupRows) == 0 {
+func (e *WindowExec) consumeGroupRows(groupRows []chunk.Row) (err error) {
+	remainingRowsInGroup := len(groupRows)
+	if remainingRowsInGroup == 0 {
 		return nil
 	}
-	e.groupRows, err = e.processor.consumeGroupRows(e.ctx, e.groupRows)
-	if err != nil {
-		return errors.Trace(err)
+	for i := 0; i < len(e.resultChunks); i++ {
+		remained := mathutil.Min(e.remainingRowsInChunk[i], remainingRowsInGroup)
+		e.remainingRowsInChunk[i] -= remained
+		remainingRowsInGroup -= remained
+
+		// TODO: Combine these three methods.
+		// The old implementation needs the processor has these three methods
+		// but now it does not have to.
+		groupRows, err = e.processor.consumeGroupRows(e.ctx, groupRows)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		_, err = e.processor.appendResult2Chunk(e.ctx, groupRows, e.resultChunks[i], remained)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if remainingRowsInGroup == 0 {
+			e.processor.resetPartialResult()
+			break
+		}
 	}
 	return nil
 }
 
-func (e *WindowExec) fetchChildIfNecessary(ctx context.Context, chk *chunk.Chunk) (err error) {
-	if e.inputIter != nil && e.inputRow != e.inputIter.End() {
-		return nil
-	}
-
-	// Before fetching a new batch of input, we should consume the last group rows.
-	err = e.consumeGroupRows()
-	if err != nil {
-		return errors.Trace(err)
+func (e *WindowExec) fetchChildIfNecessary(ctx context.Context) (EOF bool, err error) {
+	if e.inputIter != nil && e.inputIter.Current() != e.inputIter.End() {
```

Comment:
Is any possibility to get into the branch?

Reply:
When a `consumeOneGroup` exits because `meetNewGroup`, the next `consumeOneGroup` must be starts with this branch.